### DISCUSSION TITLE
A message from the creator of absurd-sql about maintenance

### DISCUSSION THREAD

**Discussion by jlongster** [jlongster/absurd-sql#33] [Announcements]
Hello! I'm opening up Discussions on absurd-sql because it's a good way to talk about the project without filling up issues.

I want to bring up something that's been weighing on me: **I've been absent**. It probably feels like I launched this project (to a lot of great interest!) and then very quickly disappeared. I'm embarrassed about it, and what's even worse it a few weeks ago I commented on several of the issues saying I was going to start picking it up again, but I didn't!

**The tl;dr is**: I'm still very excited about this project and intend to actively maintain it and be a part of relevant standards discussions. It's been difficult to find time for various reasons, but I'm going to start checking in regularly again. Eventually I'll open the repo to more maintainers to reduce the bottleneck on me.

## Some context

To give some context, I'm going to write a bit about my life right now. This is a bit of a weird place to do it, but it feels cathartic for me and I want to justify the total silence.

I launched absurd-sql August 12. I was active a for several days, but eventually I had other work to do. I didn't do much for a while, started feeling guilty, and on Sep 10 [commented](https://github.com/jlongster/absurd-sql/pull/25#issuecomment-917330489) that I was going to start going through things. Here we are on Oct 10!

Damn, time is really messed up. How was that a month ago?

I hold myself to a high (too high) bar, like I should have managed this project better, found a way to keep it growing quickly. I'm also sensitive to boldly proclaiming new things, and then never following through. I hate it when people/companies do that. In reality, open-source can be difficult and life is busy. It happens.

There's a dynamic that builds when you start feeling guilty about something: it makes it harder to confront, harder to pull up that thing and work on it. Ignoring it just makes it worse, delays it further, which brings more guilt. This happened to me here; it became something I just didn't look at because I felt too guilty about it.

I let too much guilt build up. It turned into stress that started effecting my sleep. It sounds ridiculous, but this project _is_ important to me and it never feels good to just ignore people offering help or asking questions.

I'm writing this so I can sleep well tonight :)

Here's why I've been busy the last two months:

* I pushed off a lot of work leading up to the launch. It took me a few weeks to feel caught up again, so almost immediately after launching I needed to stop working on it and catch up on everything else
* I work at Stripe, which is a demanding job, and I also manage [Actual](https://actualbudget.com) as a side project. Both these things take up almost all my time that I'm willing to spend on work (which is a lot). At Stripe we have some deadlines that required my full attentions. For night and other off-hour times, Actual has been filling it up with work that is too critical to ignore.
* I have 3 kids under 8, and in Sep school and sports starts sucking a _lot_ of time. Not to mention wanting to just be an available father and not working all the time!
* We had major renovations going on in our house which is finally finishing up.

Anyway, life has felt especially crazy over the last 2 months, and I'm really sorry I didn't even acknowledge that sooner here. What I'd like to happen is to bring on more maintainers so I'm not the bottleneck for small issues/PRs. I want to do that cautiously though and find the right people. Feel free to comment here if you are interested in helping out on a long-term basis!

**Comment by tanvirraj** [comment]
It's great you come back. it's very difficult to manage time after a full-time job, it's more difficult when you work for like Stripe. 
you also have to maintain Actual. 
One thing I'm curious about 
how you schedule your day, how many hours you have to put for Stripe 
how you manage all the time 
What's your daily routine look like

**Reply by jlongster** [reply]
I do work a lot, but not to an extreme degree. It's important to figure out how to be more productive within the same timeframe; don't just add more hours to get more stuff done. Our brains are weird and sometimes I can get just as much done in 4 hours than I would have in 3 days when I wasn't as in tune. I try to find ways to active the former state of mind so I get more done in the 40-50 hour work week. (Although to be honest I probably do work more than that)

I have kids, so I'm up by 7-8 (sometimes my wife gets up with them first). If it works out, sometimes I'll do some non-Stripe work for a bit. Around 8:45 I start getting ready for the day. Take shower, make coffee. Around 9:30 I sit down and spend an hour going through email, catching up on all communication channels, etc. If I don't have much to catch up on, some days I will work on Actual for a little bit, especially if I'm crunched for a release or something.

It varies by day, but I'll start Stripe around 9-10:30. If I started later, I end my day later. Sometime that means I'm downstairs finishing up work while the kids finish dinner. But it's all Stripe for 8-9 hours. (Lately I've been trying to work out in the middle of the day if I have a meeting that I'm just listening in on)

5-7PM is help getting the kids dinner, putting them down, etc, while maybe still finishing up work. With my wife we watch a show or hang out for a while, and I'll usually work on Actual for another hour or two at night. (This is highly variable; sometimes I'm playing through a video game instead and will do that for a couple weeks)

My schedule is highly fluid, it's hard to describe it generally. Every week feels different depending on the constraints/deadlines/etc. I'm not sometime who sticks by a routine.

I think the most important thing is to learn how to be productive for a single hour. If I have a single hour at night to work on Actual, what is the most effective thing I can do during that time? It's probably not messing with build systems, but instead sketching out a new feature as fast as I can.

The second most important thing (well, probably _the_ most important) is to watch your mental health. If I'm not inspired, all this work will burn me out. During those times I'll play video games or do other things at night instead. For me personally, it's not hard to feel inspired though which helps.

**Comment by PaulSebalu** [comment]
I'm so interested in helping out in the long term. Perhaps if you could share the ideal qualities of the right person then I can focus on being that in the meantime.

**Reply by jlongster** [reply]
Thank you! The main thing is detail-orientated and trustworthy. Need to be able to trust that if they merge a PR, they've properly reviewed it and made sure it's not going to break something unintentionally. Or if they aren't as experienced, are able to know when they should not merge a PR because they don't understand it well enough, and instead wait on me or someone else.

The hard part about this project is it's _very_ tricky. A large part of the codebase is very sensitive to changes; a one-line code change could lead to tricky deadlocks that make the whole app freeze unpredictably. So PR reviews must be very thorough, and the reviewer must understand the code really well. I'd like to add more tests, but don't think that we can rely on tests enough here because it's so tricky.

If you are interested, you can start looking at PRs and help test them. Open the PR locally, run it, and verify that the change does what it describes. You can comment on the PR with your findings. That would help a lot!

**Comment by quolpr** [comment]
Hey! Wow, so cool that you shared detailed feedback about what is happening with you/the project. I started feeling that the project began to be abandoned, but it is great that you have a plan for it. If I had a day plan as you have, I would go crazy 😄 We are human, and it happens that we can't predict how the future will go, and it's ok. Better to be healthy than to be burnt-out

Actually, I am building a Roam-like note-taking app(I am going to open-source it soon 🙂) on top of the absurd-sql, and it already works much much better than my previous solution built on top of the indexedDb. So it is fantastic what you are doing; it's like a game-changer for my project (and for the web too, I think 😄. Need more articles, use-cases to start people knowing about it). And I would be happy to help you with an absurd-sql on a long-term basis  👍

**Reply by jlongster** [reply]
Awesome, that's exciting! Looking forward to seeing your project

**Comment by rshigg** [comment]
Great to hear from you 😊

I think I speak for most of the community when I say we would rather you live a healthy, sustainable, lifestyle and get to spend valuable time with your kids. Open source is work like any other and I appreciate the work you've put into releasing this.

This project has been a game changer for me as well and I'll wait however long it takes for updates to come in since I know they'll be meaningful and effective.

Take care of you and your family first, we'll be here to help however we can.

**Reply by jlongster** [reply]
Thank you!

**Comment by rosano** [comment]
Thanks for sharing this. As with most open-source maintenance: nothing is owed. You did quite a lot by offering a compelling vision that got many people thinking, and re-thinking. Thank you.

Ideally we build the future together. Better if you're well, so take care of yourself, always.

**Comment by nicolas-albert** [comment]
Hi everyone,
Does this project continue to live somewhere or this repo can comeback to the life?

Thx!

**Reply by ivan** [reply]
You might be interested in wa-sqlite (note the license though).

**Reply by nicolas-albert** [reply]
Thanks!
It's true that GPL v3 can be annoying for a non-GPL product :-/

**Comment by tantaman** [comment]
Would you be willing to bring on a partner to help maintain `absurd-sql`?

**Comment by edmofro** [comment]
FYI, anyone coming here in 2024+
https://jlongster.com/where-have-i-been

**Reply by isaacfink** [reply]
I read the article, but I still don't know the status of this project, is it stable? Is it still maintained?

**Reply by mattbishop** [reply]
Pretty clear, the status is he has more important things to do (I agree) than drive this project. Good news, it's OSS, so if someone wants to prove themselves capable of picking up and carrying forward, fork it and start releasing your version with improvements.

**Reply by edmofro** [reply]
But also there is now an official [WASM build of sqlite](https://www.npmjs.com/package/@sqlite.org/sqlite-wasm), using OPFS - just use that

**Reply by tantaman** [reply]
there are many alternatives in 2024

- https://github.com/rhashimoto/wa-sqlite
- https://sqlite.org/wasm/doc/trunk/index.md
- https://github.com/tursodatabase/libsql-wasm-experimental
- https://github.com/vlcn-io/cr-sqlite | https://github.com/vlcn-io/js

that are maintained and have kept up with changes to the web platform. E.g., synchronous OPFS APIs that don't require COOP headers.
